lib/libzpool/coroutine.c

```diff
@@ -24,12 +24,70 @@ static __thread uzfs_coroutine_t *thread_local_coroutine = NULL;
 static timer_thread_t timer_thread;
 static uint32_t cur_key_idx = 0;
 
-#define	MAX_COROUTINE_POOL_SIZE	100
-static __thread uzfs_coroutine_t *coroutine_pool_head = NULL;
-static __thread int cur_coroutine_pool_size = 0;
+typedef struct coroutine_pool {
+	uzfs_coroutine_t *pool_head;
+	int pool_size;
+	int capacity;
+} coroutine_pool_t;
+
+static __thread coroutine_pool_t tls_coroutine_pool = {NULL, 0, 100};
+
+static pthread_mutex_t global_pool_lock = PTHREAD_MUTEX_INITIALIZER;
+static coroutine_pool_t global_coroutine_pool = {NULL, 0, 1000};
 
-#define	DEFAULT_STACK_SIZE	(1<<20)
 #define	DEFAULT_GUARD_SIZE	getpagesize()
+#define	DEFAULT_STACK_SIZE	((1<<20) - DEFAULT_GUARD_SIZE)
+
+static inline uzfs_coroutine_t *
+coroutine_pool_pop(coroutine_pool_t *pool)
+{
+	if (pool->pool_head == NULL) {
+		return (NULL);
+	}
+
+	uzfs_coroutine_t *coroutine = pool->pool_head;
+	pool->pool_head = coroutine->next_in_pool;
```

Comment:
it's better to set coroutine->next_in_pool to NULL to avoid the stale pointer

Reply:
done